Update entrypoint to use new Scrapinghub interface

chekunkov opened the pull request:
Use named pipe located in the job container to send data to Scrapinghub
backend.

chekunkov commented:
Ready for review @dangra @vshlapakov

dangra commented:
LGTM.

vshlapakov commented on sh_scrapy/extension.py:
```diff
@@ -20,14 +23,13 @@ def __init__(self, crawler):
         exporter_kwargs = {'binary': False} if not IS_PYTHON2 else {}
         with ignore_warnings(category=ScrapyDeprecationWarning):
             self.exporter = PythonItemExporter(**exporter_kwargs)
-        log.msg("HubStorage: writing items to %s" % self.hsref.job.items.url)
```
Probably makes sense to keep the log message: sometimes it's convenient to see what's the url to achieve items data, I think it's not related with changes in the PR.

Also, maybe to drop a log line saying that this sh-ep-scrapy uses pipes, or at least print its version? It could be very useful for debugging

chekunkov replied:
I removed that when I thought I'll be able to remove hsref completely, it makes sense to return it back, no problem. I can modify the message to "Writing items to %s", not mentioning pipe, will it be enough for distinction?

vshlapakov replied:
I think so, thanks

dangra replied:
I dont think that debug message is anything helpful, it was added while we were migrating from mongodb to hbase for debugging purposes much like we are migrating now to pipe.

chekunkov replied:
I can remove it again, no problem 😄

vshlapakov replied:
Ok, I didn't know all the history 👌  😄

vshlapakov commented on sh_scrapy/extension.py:
```diff
@@ -46,32 +48,32 @@ def item_scraped(self, item, spider):
-        self._seen[response] = riq
+        request_object = {
+            'url': response.url,
+            'status': int(response.status),
```
Out of curiosity, but why do you need conversion here? 
It was standardised in Scrapy more than [8 years ago](https://github.com/scrapy/scrapy/commit/f63a661320805e80b0a078752e7a4f6218141418#diff-1f09c5b98974d24900544c14f968069eR34).

chekunkov replied:
@vshlapakov Nothing prevents users from setting `response.status = '400'` or `response.status = 'blah'` in the middleware, better to be strict and safe

vshlapakov commented on sh_scrapy/extension.py:
```diff
@@ -46,32 +48,32 @@ def item_scraped(self, item, spider):
+            'fp': request_fingerprint(response.request),
+        }
+        self.pipe_writer.write_request(request_object)
+        self._seen[response] = next(self.request_id_sequence)
```
Btw, thanks for decrypting the `riq` variable, someone had to do it 😄

vshlapakov commented on sh_scrapy/stats.py:
```diff
@@ -1,30 +1,20 @@
     def _upload_stats(self):
-        row = [int(time.time() * 1000)]
-        row.extend(self._stats.get(x, 0) for x in self.STATS_TO_COLLECT)
-        self.hsref.job.samples.write(row)
```
Sorry, I don't recall all the details about samples probably, are we going to support it in the new version of ep-scrapy? I see it's supported in kumo-forwarder

chekunkov replied:
`pipe_writer.write_stats(self._stats)` writes Scrapy stats to the KF and it, in its turn, extracts and writes samples from it.

vshlapakov replied:
I see, understood 👌

vshlapakov commented:
LGTM except for a few minor things to clarify

chekunkov commented:
apologies @dangra @vshlapakov, I completely forgot about tests 🤦‍♂️. fixed existing tests and added tests for pipe writer, changed implementation a bit. ready for review.

vshlapakov commented on tests/test_writer.py:
```diff
@@ -0,0 +1,140 @@
+        w.close()
+
+
+# def test_init_fails_if_no_fifo_is_provided(tmpdir):
```
The test is commented, do we need it?

chekunkov replied:
no, I forgot to remove it, thanks :)

vshlapakov commented on tests/test_log.py:
```diff
@@ -232,12 +220,25 @@ def test_stdout_logger_writelines_empty():
+    payload = json.loads(pipe_mock.write.call_args_list[2][0][0])
+    assert isinstance(payload.pop('time'), int)
+    assert payload == {
+        'message': r'value=x\x9c+K\xcc)M\x05\x00\x06j\x02\x1e',
```
Maybe compare with `message` instead of hard-coding the value?

chekunkov replied:
@vshlapakov values are not equal: 

```
(Pdb) repr(payload['message'])
"u'value=x\\\\x9c+K\\\\xcc)M\\\\x05\\\\x00\\\\x06j\\\\x02\\\\x1e'"
(Pdb) repr(message)
"'value=x\\x9c+K\\xcc)M\\x05\\x00\\x06j\\x02\\x1e'"
```

vshlapakov commented on tests/test_writer.py:
```diff
@@ -0,0 +1,140 @@
+import os
+import threading
+
+import logging
```
Nitpicking, but let's move the import up?

vshlapakov commented:
A couple of minor notes, but overall LGTM, nicely done!

dangra commented on sh_scrapy/writer.py:
```diff
@@ -0,0 +1,66 @@
+        self._pipe.close()
+
+
+pipe_writer = _PipeWriter(os.environ['SHUB_FIFO_PATH'])
```
this will cause an import error in environments where SHUB_FIFO_PATH is not set. can it be a problem?

chekunkov replied:
@dangra this is intentional, I've noticed how hsref can have different states based on env variable, but I don't see any benefits of doing so here. We always set this env variable in the new version of scheduler, if this variable is not set the process should fail, this version of sh_scrapy is useless without this env variable set. I think benefits of failing early are clear.

chekunkov commented on sh_scrapy/writer.py:
```diff
@@ -0,0 +1,66 @@
+        self._write('FIN', {'outcome': outcome})
+
+    def flush(self):
+        self._pipe.flush()
```
I added the usage to the PipeWriter._write(self) method, should be covered now.

chekunkov replied:
But to tell the truth I'm not sure it's needed, probably have a point. If someone will decide to hack with private _pipe he should be able to call flush for _pipe, and without that it's useless, I'll remove the method.
